File: service/post_service.go
package service

import (
	"context"
	"kopherlog/domain"
	"kopherlog/repository"
)

type PostService interface {
	Write(ctx context.Context, postCreate *domain.PostCreate) error
	Get(ctx context.Context, id int) (*domain.PostResponse, error)
	GetAll(ctx context.Context, search *domain.PostSearch) ([]*domain.PostResponse, error)
	Edit(ctx context.Context, id int, edit *domain.PostEdit) error
	Delete(ctx context.Context, id int) error
}

type postService struct {
	postRepository repository.PostRepository
}

func NewPostService(postRepository repository.PostRepository) PostService {
	return &postService{postRepository: postRepository}
}

func (p *postService) Write(ctx context.Context, postCreate *domain.PostCreate) error {
	post := &domain.Post{Title: postCreate.Title, Content: postCreate.Content}
	_, err := p.postRepository.Save(ctx, post)
	if err != nil {
		return err
	}
	return nil
}

func (p *postService) Get(ctx context.Context, id int) (*domain.PostResponse, error) {
	post, err := p.postRepository.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	return &domain.PostResponse{
		ID:      post.ID,
		Title:   post.Title,
		Content: post.Content,
	}, nil
}

func (p *postService) GetAll(ctx context.Context, search *domain.PostSearch) ([]*domain.PostResponse, error) {
	posts, err := p.postRepository.FindAll(ctx, search)
	if err != nil {
		return nil, err
	}
	var postResponses []*domain.PostResponse
	for _, post := range posts {
		postResponses = append(postResponses, &domain.PostResponse{
			ID:      post.ID,
			Title:   post.Title,
			Content: post.Content,
		})
	}
	return postResponses, nil
}

func (p *postService) Edit(ctx context.Context, id int, edit *domain.PostEdit) error {
	target, err := p.postRepository.FindByID(ctx, id)
	if err != nil {
		return err
	}

	post := &domain.Post{
		Title:   coalesce(edit.Title, target.Title),
		Content: coalesce(edit.Content, target.Content),
	}
	err = p.postRepository.Update(ctx, target, post)
	if err != nil {
		return err
	}
	return nil
}

func coalesce(values ...string) string {
	for _, value := range values {
		if value != "" {
			return value
		}
	}
	return ""
}

func (p *postService) Delete(ctx context.Context, id int) error {
	_, err := p.postRepository.FindByID(ctx, id)
	if err != nil {
		return err
	}
	err = p.postRepository.Delete(ctx, id)
	if err != nil {
		return err
	}
	return nil
}

File: src/main/kotlin/jp/co/kopher/kopherlog/service/PostService.kt
package jp.co.kopher.kopherlog.service

import jp.co.kopher.kopherlog.domain.Post
import jp.co.kopher.kopherlog.exception.PostNotFound
import jp.co.kopher.kopherlog.repository.PostRepository
import jp.co.kopher.kopherlog.request.PostCreate
import jp.co.kopher.kopherlog.request.PostEdit
import jp.co.kopher.kopherlog.request.PostSearch
import jp.co.kopher.kopherlog.response.PostResponse
import org.slf4j.LoggerFactory
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class PostService(
    private val postRepository: PostRepository,
) {

    val log = LoggerFactory.getLogger(PostService::class.java)

    fun write(postCreate: PostCreate) {
        val post = Post(
            _title = postCreate.title,
            _content = postCreate.content
        )
        postRepository.save(post)
    }

    fun get(id: Long): PostResponse {
        val post = postRepository.findByIdOrNull(id) ?: throw PostNotFound()

        return PostResponse(
            id = post.id!!,
            title = post.title,
            content = post.content
        )
    }

    fun getList(search: PostSearch): List<PostResponse> {
        return postRepository.getList(search).map { PostResponse.from(it) }.toList()
    }

    @Transactional
    fun edit(id: Long, postEdit: PostEdit) {
        val post = postRepository.findByIdOrNull(id) ?: throw PostNotFound()

        post.edit(postEdit)
    }

    fun delete(id: Long) {
        val post = postRepository.findByIdOrNull(id) ?: throw PostNotFound()

        postRepository.delete(post)
    }

}